src/sagemaker_sklearn_extension/feature_extraction/sequences.py

```diff
@@ -0,0 +1,414 @@
+        else:
+            raise ValueError(
+                f"{interpolation_method} is not a supported interpolation method. Please choose one from "
+                f"the following options: [linear, fill, zeroes, hybrid]."
+            )
+        return X
+
+    @staticmethod
+    def _convert_to_df(X):
+        """Convert the list of np.arrays X into a dataframe compatible with extract_features."""
+        X_df = pd.DataFrame(data=X)
+        X_df = X_df.stack(dropna=False)
+        X_df.index.rename(["id", "time"], inplace=True)
+        X_df = X_df.reset_index()
+        return X_df
+
+    def _interpolate(self, X_df):
+        """Impute missing values through the selected interpolation_method."""
+        if self.interpolation_method is not None:
+            X_df = self._impute_ts(X_df, self.interpolation_method)
+        return X_df
+
+    @staticmethod
+    def _hybrid_interpolation(x):
+        """Replace with zeroes any NaNs at the end or start of the sequences and forward fill the remaining NaNs."""
+        # Compute the index of the first and last non-NaN value
+        # In case of all NaNs, both first_valid and last_valid are None, and all values get replaced with zeroes
+        first_valid = x[0].first_valid_index()
+        last_valid = x[0].last_valid_index()
+        x.loc[first_valid:last_valid, 0] = x.loc[first_valid:last_valid, 0].interpolate(method="ffill", axis=0)
+        x[0] = x[0].fillna(0)
+        return x
+
+    def _extract_tsfresh_features(self, X):
+        X_df = self._convert_to_df(X)
+        X_df_no_nans = X_df.dropna()
+        if self.extraction_type == "minimal":
+            extraction_setting = MinimalFCParameters()
+        elif self.extraction_type == "efficient":
+            extraction_setting = EfficientFCParameters()
+        elif self.extraction_type == "all":
+            extraction_setting = ComprehensiveFCParameters()
+        else:
+            raise ValueError(
+                f"{self.extraction_type} is not a supported feature extraction option. Please choose one from "
+                f"the following options: [minimal, efficient, all]."
+            )
+        # Extract time series features from the dataframe
+        # Replace any ``NaNs`` and ``infs`` in the extracted features with median/extreme values for that column
+        tsfresh_features = extract_features(
+            X_df_no_nans,
+            default_fc_parameters=extraction_setting,
+            column_id="id",
+            column_sort="time",
+            impute_function=impute,
+        )
+        # If X_df.dropna() dropped some observations entirely (i.e., due to all NaNs),
+        # impute each tsfresh feature for those observations with the median of that tsfresh feature
+        tsfresh_features_imputed = impute(tsfresh_features.reindex(pd.RangeIndex(X_df["id"].max() + 1)))
+        return tsfresh_features_imputed, X_df
+
+    def _more_tags(self):
+        return {"_skip_test": True, "allow_nan": True}
```

Comment:
Why is `_skip_test` set to True?

Reply:
Without _skip_test set to True, test_common.py would fail since this estimator takes as input a list of possibly variable-length np.arrays, which is not a type supported by those tests.

Reply:
How can np.arrays be variable length? I thought our input is usually a 1D numpy array of strings, where the strings are sequences of variable length, but the input array will always be of fixed length.

Reply:
That's the case for TSFeatureExtractor and TSFlattener, whose input is an array of strings. However, this transformer (TSFreshFeatureExtractor) takes as input the numerical sequences already flattened by TSFlattener, which can be variable length.